File: src/problems/leetcode-cn/22. 括号生成/solution/java/Solution.java
class Solution {
    List<String> result = new LinkedList<>();

    public List<String> generateParenthesis(int n) {
        backtrack("", 0, 0, n);
        return result;
    }

    public void backtrack(String path, int open, int close, int max) {
        if (path.length() == 2*max) {
            result.add(path);
            return;
        }

        if (open < max) {
            backtrack(path+"(", open+1, close, max);
        }
        if (close < open) {
            backtrack(path+")", open, close+1, max);
        }
    }
}

File: src/problems/leetcode-cn/22. 括号生成/solution/go/solution.go
func generateParenthesis(n int) []string {
    result := []string{}

    var backtrack func(cur string, open, close, max int)

    backtrack = func(cur string, open, close, max int) {
        if len(cur) == max*2 {
            result = append(result, cur)
        }
        if open < max {
            backtrack(cur+"(", open+1, close, max)
        }
        if close < open {
            backtrack(cur+")", open, close+1, max)
        }
    }

    backtrack("", 0, 0, n)

    return result
}